tcms/testruns/static/testruns/js/get.js

```diff
@@ -1,5 +1,88 @@
 
+    const testRunId = $('#test_run_pk').data('pk')
+
+    jsonRPC('TestExecutionStatus.filter', {}, data => {
+        executionStatuses = data
+        drawPercentBar(testRunId)
+    })
 })
+
+function drawPercentBar(testRunId) {
+
+    jsonRPC('TestExecution.filter', { 'run_id': testRunId }, testExecutions => {
+
+        let positiveCount = 0;
+        let negativeCount = 0;
+        let allCount = testExecutions.length;
+        let statusCount = {}
+        executionStatuses.forEach(s => statusCount[s.name] = { count: 0, id: s.id })
+
+        testExecutions.forEach(testExecution => {
+            const executionStatus = executionStatuses.find(s => s.id === testExecution.status_id)
+
+            if (executionStatus.weight > 0) {
+                positiveCount++
+            } else if (executionStatus.weight < 0) {
+                negativeCount++
+            }
+
+            statusCount[executionStatus.name].count++
+        })
+
+        renderProgressBars(positiveCount, negativeCount, allCount)
+        renderCountPerStatusList(statusCount)
+    })
+}
+
+function renderProgressBars(positiveCount, negativeCount, allCount) {
+
+    const positivePercent = +(positiveCount / allCount * 100).toFixed(2)
+    const positiveBar = $(".progress > .progress-completed")
+    if (positivePercent) {
+        positiveBar.text(`${positivePercent}%`)
+    }
+    positiveBar.css('width', `${positivePercent}%`)
+    positiveBar.attr('title', `${positivePercent}% Completed`)
+    positiveBar.attr('aria-valuenow', `${positivePercent}`)
+
+    const negativePercent = +(negativeCount / allCount * 100).toFixed(2)
+    const negativeBar = $('.progress > .progress-failed')
+    if (negativePercent) {
+        negativeBar.text(`${negativePercent}%`)
+    }
+    negativeBar.css('width', `${negativePercent}%`)
+    negativeBar.attr('title', `${negativePercent}% Failed`)
+    negativeBar.attr('aria-valuenow', `${negativePercent}`)
+
+    const remainingPercent = 100 - (negativePercent + positivePercent)
+    const remainingBar = $('.progress > .progress-bar-remaining')
+    if (remainingPercent) {
+        remainingBar.text(`${remainingPercent}%`)
+    }
+    remainingBar.css('width', `${remainingPercent}%`)
+    remainingBar.attr('title', `${remainingPercent}% Remaining`)
+    remainingBar.attr('aria-valuenow', `${remainingPercent}`)
+
+    $(".total-execution-count").text(allCount)
+}
+
+function renderCountPerStatusList(statusCount) {
+    $(".count-per-status-list").remove()
+    $(".count-per-status-container").prepend('<span class="count-per-status-list"></span>')
+
+    for (var status in statusCount) {
+        const count = statusCount[status].count
+        const element = count ? `<a href="?status_id=${statusCount[status].id}">${count}</a>` : '0'
+
+        $(".count-per-status-list").append(`
+      <li class="list-group-item" style="padding: 0;">
+        <label>${status}</label> - ${element}
+      </li>
+      `)
```

Comment:
I don't like rendering HTML inside JavaScript. The HTML template can render all status elements with their labels and count 0. It can also add data attribute or ID so these elements are easier to select via JS and then `renderCountPerStatusList` will only update the counts foe every list item, possibly only for 1 li based on the status ID (can be used later when updating TEs).

Reply:
'HTML in JS' is used in ReactJS and is becoming quite standard thing, so it is not something weird. Also, it is the better approach for something that is going to be changed dynamically (e.g. when a button is clicked, or an event is triggered). Because you just delete the whole thing and render it again. Also, the new JS template strings makes in easy and neat. 

IMO this is way better than selecting multiple elements and editing them one by one, because:
1. DOM operations are slow. Deleting part of the DOM and inserting something new in its place is faster than selecting 10 elements by ID and changing them value one by one, because these are 10 DOM manipulations (vs 1 in the other approach)(and this is the way React works - it has a virtual DOM in memory, and even if you edit 10 elements, under the hood it groups them and in the end applies the changes as 1 DOM manipulation).
2. It's one thing to select elements by IDs like `status-button` or `status-bar` and another to have to build IDs once in the HTML templates and another time in the JS. It's just ugly and will quickly get out of hand.

Reply:
> HTML in JS' is used in ReactJS and is becoming quite standard thing

this is not a reason to use something. 

> 1. DOM operations are slow.

with all the status li's rendered via the templating layer you are not deleting DOM elements, just updating their innerHTML attribute to the correct number value. This is something that is done anyway upon update of test executions to refresh the UI.  Much better to have a minimalist 1 line update (or a small function if you like) and use it in several places instead of rendering <li> tags whenever this card needs to be updated.

>2. It's one thing to select elements by IDs like status-button or status-bar and ...

```
$(`#status-${statusCount[status].id`).html(count)
```

That's hardly getting out of hand and definitely less ugly than building the HTML contents in 2 different pieces of code.

Also with the current implementation the labels with value 0 are not clickable which introduces unnecessary login on the FE. All of them should be exactly the same minus the text content.

Reply:
> this is not a reason to use something. 

sure, it isn't. but 'I don't like something' isn't one as well as that is subjective. My point was that this is not something weird, nor something custom that only we are doing.

> definitely less ugly than building the HTML 

that is subjective. my opinion differs.

> Much better to have a minimalist 1 line update (or a small function if you like) and use it in several places instead of rendering 

again, this is subjection. I don't see anything bad with rendering.

> Also with the current implementation the labels with value 0 are not clickable which introduces unnecessary login on the FE

I don't see how this concern this discussion. if the behavior should be different, it can be achieved both ways.

Reply:
> The HTML template can render all status elements with their labels and count 0.

No, it can't, because since https://github.com/kiwitcms/Kiwi/commit/6189ad4bb01c5ec9ba537819bea68b1a77b388e6
the additional data that was sent by the view is gone.

Also, this is not new code, it's just copied from https://github.com/kiwitcms/Kiwi/commit/1dd0dacab167ede4f91a5be9af14a8a5651e7e71#diff-f1f966298171ba7b8ab6c743bd88e634R738-R745

So, changing the approach would make the things more ugly in both the template, js, and the already simplified view.